solutions/kubernetes/ceph-binary-build.sh

```diff
@@ -0,0 +1,299 @@
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For any questions about this software or licensing,
+# please email opensource@seagate.com or cortx-questions@seagate.com.
+#
+
+######################################################################
+# This script is specific for the reserved HW by RE team.
+# To use on other HW/VM please install docker and change
+# container volume $MOUNT_LOCATION for build os container.
+######################################################################
+
+source functions.sh
+
+MOUNT_LOCATION="/var/log/ceph-build"
+
+ACTION="$1"
+if [ -z "$ACTION" ]; then
+    echo "ERROR : No option provided"
+    usage
+    exit 1
+fi
+
+function usage() {
+    cat << HEREDOC
+Usage : $0 [--build-ubuntu, --build-centos, --build-rockylinux]
+where,
+    --build-ubuntu - Build Ubuntu binary packages (*.debs).
+    --build-centos  - Build CentOS binary packages (*.rpms).
+    --build-rockylinux - Build Rocky Linux binary packages (*.rpms).
+HEREDOC
+}
+
+function check_params() {
+    if [ -z "$CEPH_REPO" ]; then echo "CEPH_REPO not provided. Using default: ceph/ceph ";CEPH_REPO="ceph/ceph"; fi
+    if [ -z "$CEPH_BRANCH" ]; then echo "CEPH_BRANCH not provided. Using default: quincy";CEPH_BRANCH="quincy"; fi
+
+   echo -e "\n\n########################################################################"
+   echo -e "# CEPH_REPO         : $CEPH_REPO                  "
+   echo -e "# CEPH_BRANCH       : $CEPH_BRANCH                "
+   echo -e "#########################################################################"
+}
+
+function prereq() {
+    add_primary_separator "\t\tPreparing Build node"
+
+    # Verify docker installation:
+    add_secondary_separator "Verify docker installation"
+    if ! which docker; then
+        add_secondary_separator "Please install Docker on Build Node Agent"
+        exit 1
+    fi
+}
+
+function build_ubuntu() {
+    add_primary_separator "Building ubuntu binary packages"
+
+    add_secondary_separator "Verfiy/pull Ubuntu docker image"
+    if [[ $(docker images | tail -n+2 | awk '{ print $1":"$2 }' | grep ubuntu) != "ubuntu:latest" ]]; then
+        docker pull ubuntu:latest
+    fi
+
+    add_secondary_separator "Make build directory and run Ubuntu container"
+    mkdir -p $MOUNT_LOCATION/ubuntu
+    docker run --rm -t -d --name ceph_ubuntu -v /$MOUNT_LOCATION/ubuntu:/home ubuntu:latest
+
+    pushd $MOUNT_LOCATION/ubuntu
+        add_common_separator "Removing previous files"
+        rm -rvf *
+    popd
+
+    add_secondary_separator "Write build script"
+    cp functions.sh $MOUNT_LOCATION/ubuntu
+    pushd $MOUNT_LOCATION/ubuntu
+        cat << EOF > build.sh
```

Comment:
why are we creating script file on the go and not creating it dynamic to accept OS and work?

Reply:
- The script will run on specific container for that build os.
- Its compute intensive process hence only no concurrent builds.
- The build outputs are volume mounted to containers and output to different dirs.
- The steps are different for each os hence instead of cramming all in one build function separated them as per os.
- For less dependencies and more flexibility I am writing build steps scripts in the location mounted to home on container, then executing the build.sh script. This avoids requirements for docker plugin for jenkins and making the image with the build.sh script...the approach is similar to docker cp.
- Build container will be ephemeral and independent of hw/vm but rpms/debs will probably be copied to nfs.
- Current dependency is on MOUNT_POINT which I kept as /var/log as it comes with most space by default on all hw systems. But in future we can move to any vm just modifying the MOUNT_POINT.

- One additional thing remaining is that job should fail on script fail...just exploring that.

We can also explore other approaches for build process.